In [1]:
import csv

f = open('guns.csv')
data = list(csv.reader(f))
data[:5]

[['',
  'year',
  'month',
  'intent',
  'police',
  'sex',
  'age',
  'race',
  'hispanic',
  'place',
  'education'],
 ['1',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '34',
  'Asian/Pacific Islander',
  '100',
  'Home',
  '4'],
 ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'],
 ['3',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '60',
  'White',
  '100',
  'Other specified',
  '4'],
 ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4']]

In [2]:
headers = data.pop(0)
headers

['',
 'year',
 'month',
 'intent',
 'police',
 'sex',
 'age',
 'race',
 'hispanic',
 'place',
 'education']

In [3]:
data[:5]

[['1',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '34',
  'Asian/Pacific Islander',
  '100',
  'Home',
  '4'],
 ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'],
 ['3',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '60',
  'White',
  '100',
  'Other specified',
  '4'],
 ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4'],
 ['5',
  '2012',
  '02',
  'Suicide',
  '0',
  'M',
  '31',
  'White',
  '100',
  'Other specified',
  '2']]

In [6]:
years = []

for row in data:
    years.append(row[1])
years

['2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 '2012',
 

In [11]:
year_counts = {}

for y in years:
    if y not in year_counts:
        year_counts[y] = 1
    else:
        year_counts[y] += 1
year_counts

{'2012': 33563, '2013': 33636, '2014': 33599}

In [13]:
import datetime 

dates = []
for row in data:
    date = datetime.datetime(year= int(row[1]), month = int(row[2]), day = 1)
    dates.append(date)
dates[:5]
    


[datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 1, 1, 0, 0),
 datetime.datetime(2012, 2, 1, 0, 0),
 datetime.datetime(2012, 2, 1, 0, 0)]

In [14]:
date_counts = {}

for date in dates:
    if date not in date_counts:
        date_counts[date] = 1
    elif date in date_counts:
        date_counts[date] += 1

date_counts

{datetime.datetime(2012, 1, 1, 0, 0): 2758,
 datetime.datetime(2012, 2, 1, 0, 0): 2357,
 datetime.datetime(2012, 3, 1, 0, 0): 2743,
 datetime.datetime(2012, 4, 1, 0, 0): 2795,
 datetime.datetime(2012, 5, 1, 0, 0): 2999,
 datetime.datetime(2012, 6, 1, 0, 0): 2826,
 datetime.datetime(2012, 7, 1, 0, 0): 3026,
 datetime.datetime(2012, 8, 1, 0, 0): 2954,
 datetime.datetime(2012, 9, 1, 0, 0): 2852,
 datetime.datetime(2012, 10, 1, 0, 0): 2733,
 datetime.datetime(2012, 11, 1, 0, 0): 2729,
 datetime.datetime(2012, 12, 1, 0, 0): 2791,
 datetime.datetime(2013, 1, 1, 0, 0): 2864,
 datetime.datetime(2013, 2, 1, 0, 0): 2375,
 datetime.datetime(2013, 3, 1, 0, 0): 2862,
 datetime.datetime(2013, 4, 1, 0, 0): 2798,
 datetime.datetime(2013, 5, 1, 0, 0): 2806,
 datetime.datetime(2013, 6, 1, 0, 0): 2920,
 datetime.datetime(2013, 7, 1, 0, 0): 3079,
 datetime.datetime(2013, 8, 1, 0, 0): 2859,
 datetime.datetime(2013, 9, 1, 0, 0): 2742,
 datetime.datetime(2013, 10, 1, 0, 0): 2808,
 datetime.datetime(2013, 11,

In [17]:
def column_count(column_number):
    colValues = []
    for row in data:
        colValues.append(row[column_number])

    colCounts = {}

    for value in colValues:
        if value not in colCounts:
            colCounts[value] = 1
        elif value in colCounts:
            colCounts[value] += 1
    return colCounts

In [18]:
sex_counts = column_count(5)

In [19]:
sex_counts

{'F': 14449, 'M': 86349}

In [20]:
race_counts = column_count(7)

In [21]:
race_counts

{'Asian/Pacific Islander': 1326,
 'Black': 23296,
 'Hispanic': 9022,
 'Native American/Native Alaskan': 917,
 'White': 66237}

# Sex: ~6 x more M
# Race: Predominantly white. 
For both, what happens when I normalize value to fraction of each category in population?
police invovement in race

In [34]:
census = list(csv.reader(open('census.csv', 'r')))
census
census[0][-7]

'Race Alone - White'

In [41]:
#dict that maps each key from race_counts to pop race count in census
mapping = {
    'Asian/Pacific Islander': int(census[1][-2]) + int(census[1][-3]),
    'Black': int(census[1][-5]),
    'Native American/Native Alaskan': int(census[1][-4]),
    'Hispanic': int(census[1][-6]),
    'White': int(census[1][-7])
}
mapping

{'Asian/Pacific Islander': 15834141,
 'Black': 40250635,
 'Hispanic': 44618105,
 'Native American/Native Alaskan': 3739506,
 'White': 197318956}

In [42]:
race_per_hundredk = {}

for race in race_counts:
    race_per_hundredk[race] = (race_counts[race] / mapping[race]) * 100000
race_per_hundredk
    
    

{'Asian/Pacific Islander': 8.374309664161762,
 'Black': 57.8773477735196,
 'Hispanic': 20.220491210910907,
 'Native American/Native Alaskan': 24.521955573811088,
 'White': 33.56849303419181}

In [50]:
intents = []
for row in data:
    intents.append(row[3])

races = []
for row in data:
    races.append(row[7])

homicide_race_per_hundredk = {}

for i, race in enumerate(races):
    if intents[i] == 'Homicide':
        if race not in homicide_race_per_hundredk:
            homicide_race_per_hundredk[race] = 1
        elif race in homicide_race_per_hundredk:
            homicide_race_per_hundredk[race] += 1
            
for race in homicide_race_per_hundredk:
    homicide_race_per_hundredk[race] = 100000 * (homicide_race_per_hundredk[race]/ mapping[race])

homicide_race_per_hundredk

{'Asian/Pacific Islander': 3.530346230970155,
 'Black': 48.471284987180944,
 'Hispanic': 12.627161104219914,
 'Native American/Native Alaskan': 8.717729026240365,
 'White': 4.6356417981453335}

Homicide per 100,000 individuals of race):
{'Asian/Pacific Islander': 3.530346230970155,
 'Black': 48.471284987180944,
 'Hispanic': 12.627161104219914,
 'Native American/Native Alaskan': 8.717729026240365,
 'White': 4.6356417981453335}
 
Further: Police involvement in homicides by race